source/bayes_simulation.Rmd

```diff
@@ -45,112 +45,283 @@ correctly identifies 80 percent as "OK" but says that 20 percent are
-to be "OK"
+Fraction of faulty cars out of approved cars: 0.03 / (0.03 + 0.56) = 0.0508
 
-"31 - 86" = OK, "OK"
+**Calculation using simulation:**
 
-"87 - 100" = OK, "faulty"
+1. We note that we are only interested in outcomes where the mechanic says the cars are OK.
 
-`           REPEAT 1000         `
+2. We start with N cars.  For each car, generate a label of either
+   "faulty" or "working" with 0.3 and 0.7 probabilities respectively.
 
-Do 1000 repetitions
+3. For each faulty car, we generate one of two labels, "mechanic
+   approved" or "mechanic failed", with probabilities of 0.1 or 0.9
+   respectively.
 
-`           GENERATE 1 1,100 a         `
+4. For each working car, generate one of two labels, "mechanic
+   approved" or "mechanic failed", with probabilities 0.7 and 0.3
+   respectively.
 
-Generate a number between "1" and "100"
+5. Out of all cars "mechanic approved", count how many are "faulty".
+   The ratio between these numbers is our answer.
 
-`           IF a between 28 86         `
+Here is the code:
 
-If it's between "28" and "86" (those that say "good")
+```{python}
+import numpy as np
 
-`           SCORE a z         `
+N = 100000  # number of cars
 
-Score this number
+# Generate an array with as many entries as there are cars, each
+# being either 'working' or 'faulty'
+cars = np.random.choice(['working', 'faulty'], p=[0.7, 0.3], size=N)
 
-`           END         `
+# Count how many cars are working and how many are faulty
+N_working = np.count_nonzero(cars == 'working')
+N_faulty = N - N_working
 
-End the IF condition
+# Create a new array in which to store mechanic approved (True)
+# or mechanic declared faulty (False)
+approved = np.zeros_like(cars, dtype=bool)
 
-`           END         `
+# For each working car, generate a mechanic approved/declared faulty
+# label with the given probabilities
+approved[cars == 'working'] = np.random.choice(
+    [True, False], p=[0.8, 0.2], size=N_working
+)
 
-End REPEAT loop
+# For each faulty car, generate a mechanic approved/declared faulty
+# label with the given probabilities
+approved[cars == 'faulty'] = np.random.choice(
+    [True, False], p=[0.1, 0.9], size=N_faulty
+)
 
-`           COUNT z between 28 30 k         `
+# Select out only the cars that are faulty and were approved
+faulty_but_approved = np.logical_and(cars == 'faulty', approved)
 
-How many of the SCORED numbers were between "28\--30" (faulty, "OK")
+# Count the number of cars that are faulty but approved, as well as
+# the total number of cars that were approved
+N_faulty_but_approved = np.count_nonzero(faulty_but_approved)
+N_approved = np.count_nonzero(approved)
 
-`           SIZE z s         `
+# Calculate the ratio, which is the answer we seek
+k = N_faulty_but_approved / N_approved
 
-How many numbers were scored
+print(f'{k * 100:.2}%')
+```
 
-`           DIVIDE k s kk         `
+In this experiment, we use 100000 cars, and the results are very
```

Comment:
Yeah, now I see what you mean.  This is quite a deep question about the population and the sample.  Normally in statistics we assume the population is very large, and the sample is much smaller, so the question for simulation is just whether the number of samples we have taken (say 10000) is large enough to get a good estimate of the parameter we want - such as a probability.   We discussed this just now...